internal/kinds/capp-revision/controllers/revision_controller.go

```diff
@@ -0,0 +1,105 @@
+	logger.Info("Starting Reconcile")
+	capp := cappv1alpha1.Capp{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &capp); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info(fmt.Sprintf("Didn't find Capp: %s, from the namespace: %s", capp.Name, capp.Namespace))
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get Capp: %s", err.Error())
+	}
+	if !capp.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	if err := syncCappRevision(ctx, r.Client, capp, logger); err != nil {
+		if errors.IsConflict(err) {
+			logger.Info(fmt.Sprintf("Conflict detected requeuing: %s", err.Error()))
+			return ctrl.Result{RequeueAfter: RequeueTime}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to sync Capp: %s", err.Error())
+	}
+	return ctrl.Result{}, nil
+}
+
+// syncCappRevision manages the lifecycle of CappRevisions based on the state of a Capp, handling creation, update, or deletion.
+func syncCappRevision(ctx context.Context, k8sClient client.Client, capp cappv1alpha1.Capp, logger logr.Logger) error {
+	cappRevisions, err := adapters.GetCappRevisions(ctx, k8sClient, capp)
+	if err != nil {
+		logger.Error(err, "could not fetch cappRevisions")
+		return err
+	}
+
+	if len(cappRevisions) == 0 && capp.DeletionTimestamp.IsZero() {
+		return action_mangers.HandleCappCreation(ctx, k8sClient, capp, logger)
+		//} else if !capp.DeletionTimestamp.IsZero() {
+		//	return action_mangers.HandleCappDeletion(ctx, k8sClient, capp, logger, cappRevisions)
+	}
+	return action_mangers.HandleCappUpdate(ctx, k8sClient, capp, logger, cappRevisions)
+
+}
+
+type SpecChangedPredicate struct {
+	predicate.Funcs
+}
+
+// Update implements default UpdateEvent filter for checking label change.
+func (SpecChangedPredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil {
+		return false
+	}
+	if e.ObjectNew == nil {
+		return false
+	}
+	newCapp := e.ObjectNew.(*cappv1alpha1.Capp)
+	oldCapp := e.ObjectOld.(*cappv1alpha1.Capp)
+	return !reflect.DeepEqual(oldCapp.Spec, newCapp.Spec)
+
+}
+
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capps,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capps/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capps/finalizers,verbs=update
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capprevisions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capprevisions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=rcs.dana.io,resources=capprevisions/finalizers,verbs=update
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CappRevisionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&cappv1alpha1.Capp{}).
+		WithEventFilter(predicate.Or(predicate.AnnotationChangedPredicate{}, predicate.LabelChangedPredicate{}, SpecChangedPredicate{})).
+		Complete(r)
+}
```

Comment:
Move this entire section to the top of the file. I think it's easier when you first open a Reconciler that the `SetupWithManager` is at the top so you understand what events trigger it